rust/gsc/src/main.rs

```diff
@@ -1,27 +1,320 @@
+            hasher.update(&file.path.len().to_le_bytes());
+            hasher.update(file.path.as_bytes());
+            hasher.update(&file.hash.len().to_le_bytes());
+            hasher.update(&file.hash);
+        }
+        hasher.finalize().into()
+    }
+}
+
+#[derive(Debug, Deserialize)]
+struct ReadJobPayload {
+    cmd: String,
+    env: String,
+    cwd: String,
+    stdin: String,
+    is_atty: bool,
+    #[serde(with = "serde_bytes")]
+    hidden_info: Vec<u8>,
+    visible_files: Vec<VisibleFile>,
+}
+
+impl ReadJobPayload {
+    // TODO: Figure out a way to de-dup this with AddJobPayload somehow
+    fn hash(&self) -> [u8; 32] {
+        let mut hasher = blake3::Hasher::new();
+        hasher.update(&self.cmd.len().to_le_bytes());
+        hasher.update(self.cmd.as_bytes());
+        hasher.update(&self.env.len().to_le_bytes());
+        hasher.update(self.env.as_bytes());
+        hasher.update(&self.cwd.len().to_le_bytes());
+        hasher.update(self.cwd.as_bytes());
+        hasher.update(&self.stdin.len().to_le_bytes());
+        hasher.update(self.stdin.as_bytes());
+        hasher.update(&self.hidden_info.len().to_le_bytes());
+        hasher.update(self.hidden_info.as_slice());
+        hasher.update(&[self.is_atty as u8]);
+        hasher.update(&self.visible_files.len().to_le_bytes());
+        for file in &self.visible_files {
+            hasher.update(&file.path.len().to_le_bytes());
+            hasher.update(file.path.as_bytes());
+            hasher.update(&file.hash.len().to_le_bytes());
+            hasher.update(&file.hash);
+        }
+        hasher.finalize().into()
+    }
+}
+
+#[derive(Debug, Serialize, Deserialize)]
+#[serde(tag = "type")]
+enum ReadJobResponse {
+    NoMatch,
+    Match {
+        output_symlinks: Vec<Symlink>,
+        output_dirs: Vec<Dir>,
+        output_files: Vec<File>,
+        #[serde(with = "serde_bytes")]
+        stdout: Vec<u8>,
+        #[serde(with = "serde_bytes")]
+        stderr: Vec<u8>,
+        status: i32,
+        runtime: f64,
+        cputime: f64,
+        memory: u64,
+        ibytes: u64,
+        obytes: u64,
+    },
+}
+
+#[tracing::instrument]
+async fn read_job(
+    Json(payload): Json<ReadJobPayload>,
+    conn: Arc<DatabaseConnection>,
+) -> Json<ReadJobResponse> {
+    // First find the hash so we can look up the exact job
+    let hash: Vec<u8> = payload.hash().into();
+
+    let result = conn
+        .as_ref()
+        .transaction::<_, ReadJobResponse, DbErr>(|txn| {
+            Box::pin(async move {
+                let Some(matching_job) = job::Entity::find()
+                    .filter(job::Column::Hash.eq(hash))
+                    .one(txn)
+                    .await?
+                else {
+                  return Ok(ReadJobResponse::NoMatch);
+                };
+
+                let output_files = output_file::Entity::find()
+                    .filter(output_file::Column::JobId.eq(matching_job.id))
+                    .all(txn)
```

Comment:
I'll give a shot!

Reply:
ok so this is slightly more complicated it turns out. We spoke offline about it but I'll leave it here for later. TL;DR I was able to use lazy-loading via `find_related` to clean up the code a bit but the performance will not be different (Should be the same queries even). The reason that eager-loading is a bad idea is because it will duplicate the stdin/stdout which could be quite large. There are however ways to efficently do this in 2 qurries but its a bit of a headache so we don't want to do that right now. There are much lower hanging performance gains to be made when we decide to optimize.